src/sentry/notifications/notificationcontroller.py

```diff
@@ -0,0 +1,460 @@
+        return project_settings | org_settings | team_or_user_settings
+
+    def _get_scope_for_layering(
+        self, recipient: RpcActor | Team | RpcUser
+    ) -> Tuple[NotificationScopeEnum, Iterable[int | RpcActor | Team | RpcUser]]:
+        if self.project_ids:
+            return (NotificationScopeEnum.PROJECT, self.project_ids)
+        if self.organization_id is not None:
+            return (NotificationScopeEnum.ORGANIZATION, [self.organization_id])
+        if recipient_is_user(recipient):
+            return (NotificationScopeEnum.USER, [recipient])
+        if recipient_is_team(recipient):
+            return (NotificationScopeEnum.TEAM, [recipient])
+
+        raise ValueError(f"Recipient is of unsupported type: {recipient}")
+
+    def _filter_options(
+        self,
+        settings: Iterable[NotificationSettingOption],
+        **kwargs,
+    ) -> list[NotificationSettingOption]:
+        return [
+            setting
+            for setting in settings
+            if all(getattr(setting, arg) == kwargs[arg] for arg in kwargs)
+        ]
+
+    def _filter_providers(
+        self,
+        settings: Iterable[NotificationSettingProvider],
+        **kwargs,
+    ) -> list[NotificationSettingProvider]:
+        return [
+            setting
+            for setting in settings
+            if all(getattr(setting, arg) == kwargs[arg] for arg in kwargs)
+        ]
+
+    def _get_layered_setting_options(
+        self,
+        **kwargs,
+    ) -> MutableMapping[
+        Recipient,
+        MutableMapping[
+            Scope, MutableMapping[NotificationSettingEnum, NotificationSettingsOptionEnum]
+        ],
+    ]:
+        """
+        Returns a mapping of the most specific notification setting options for the given recipients and scopes.
+        Note that this includes default settings for any notification types that are not set.
+
+        Args:
+            setting_type: If specified, only return settings of this type.
+        """
+        settings = self.get_all_setting_options()
+        project_settings = self._filter_options(
+            settings=settings, scope_type=NotificationScopeEnum.PROJECT.value, **kwargs
+        )
+        org_settings = self._filter_options(
+            settings=settings, scope_type=NotificationScopeEnum.ORGANIZATION.value, **kwargs
+        )
+        user_settings = self._filter_options(
+            settings=settings, scope_type=NotificationScopeEnum.USER.value, **kwargs
+        )
+        scoped_settings = [project_settings, org_settings, user_settings]
+
+        defaults = get_type_defaults()
+        layered_setting_options: MutableMapping[
+            Recipient,
+            MutableMapping[
+                Scope, MutableMapping[NotificationSettingEnum, NotificationSettingsOptionEnum]
+            ],
+        ] = defaultdict(lambda: defaultdict(dict))
+        for recipient in self.recipients:
+            (scope_type, scope_items) = self._get_scope_for_layering(recipient)
```

Comment:
@snigdhas does this handle the case where if there is a project setting we also need to check the organization setting? The UI doesn't support this but I think the backend should.

Reply:
Yes, it should -- regardles off the `scope_type`, we still check project | organization | user settings in L189-L200. We'll iterate over the three setting levels in `scoped_settings` until we find a valid setting, or resort to the default instead.